unicore/distribute/api/repos.py

```diff
@@ -73,7 +92,85 @@ def post(self):
                     'url': self.request.route_url('repositoryresource',
                                                   name=name)
                 }))
-        return format_diffindex(changes)
+        return list(format_diffindex(changes))
+
+
+def initialize_repo_index(event):
+    repo = event.repo
+    model_mappings = map(
+        lambda (content_type, mapping): (load_model_class(repo, content_type),
+                                         mapping),
+        event.mapping.iteritems())
+    sm = StorageManager(repo)
+    im = ESManager(
+        storage_manager=sm,
+        es=get_es(event.config),
+        index_prefix=get_index_prefix(repo.working_dir))
+
+    if im.index_exists(sm.active_branch()):
+        im.destroy_index(sm.active_branch())
+    im.create_index(sm.active_branch())
+
+    try:
+        for model_class, mapping in model_mappings:
+            if mapping is None:
+                im.setup_mapping(sm.active_branch(), model_class)
+            else:
+                im.setup_custom_mapping(
+                    sm.active_branch(), model_class, mapping)
+
+        for model_class, _ in model_mappings:
+            for model in sm.iterate(model_class):
+                im.index(model)
+    except ElasticsearchException:
+        im.destroy_index(sm.active_branch())
+        raise
+
+
+def update_repo_index(event):
+    repo = event.repo
+    changes = event.changes
+    workspace = Workspace(
+        repo=repo,
+        es={'urls': [event.config.get('es.host', 'http://localhost:9200')]},
+        index_prefix=get_index_prefix(repo.working_dir))
+    sm = workspace.sm
+    im = workspace.im
+
+    old_branch = repo.head.ref
+    repo.heads[event.branch].checkout()
+
+    if len(repo.remotes) > 1 and any(changes):
+        workspace.reindex_diff(changes)
+        return
+
+    if any(changes.iter_change_type('R')):
+        workspace.reindex_diff(changes)
+        return
+
+    # unindex deleted blobs
+    for diff in changes.iter_change_type('D'):
+        path_info = sm.path_info(diff.a_blob.path)
+        if path_info is None:
+            continue
+        im.raw_unindex(*path_info)
+
+    # reindex added blobs
+    for diff in changes.iter_change_type('A'):
+        path_info = sm.path_info(diff.b_blob.path)
+        if path_info is None:
+            continue
+        obj = sm.get(*path_info)
+        im.index(obj)
+
+    # reindex modified blobs
+    for diff in changes.iter_change_type('M'):
+        path_info = sm.path_info(diff.a_blob.path)
+        if path_info is None:
+            continue
+        obj = sm.get(*path_info)
```

Comment:
This looks like it was copied from elastic-git's local workspace, is there a way we can prevent the duplication since we already have the `Workspace` object here and this code mostly lives in the `Workspace.pull` method?